# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/cookiestore/CookieStore.cpp
# Commit: 30d5dfedf252
# Full Hash: 30d5dfedf252f4f10a0d8948e86f4e57f1c4ccf9
# Author: Andrea Marchesini <amarchesini@mozilla.com>
# Date: 2024-09-11 09:21:34
# Regressor Bug: 1475599
# File Overlap Count: 1
# Description:
#   Bug 1475599 - part 8 - CookieStore API - e10s support, r=smaug,cookie-reviewers,timhuang
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D217574
# ==============================================================================

diff -r fc3b4c5ee4b2 -r 30d5dfedf252 dom/cookiestore/CookieStore.cpp
--- a/dom/cookiestore/CookieStore.cpp	Tue Sep 10 16:31:40 2024 +0000
+++ b/dom/cookiestore/CookieStore.cpp	Tue Sep 10 16:31:41 2024 +0000
@@ -7,6 +7,7 @@
 #include "CookieStore.h"
 #include "CookieStoreChild.h"
 #include "CookieStoreNotifier.h"
+#include "CookieStoreNotificationWatcherWrapper.h"
 
 #include "mozilla/dom/Document.h"
 #include "mozilla/dom/Promise.h"
@@ -219,6 +220,10 @@
 CookieStore::CookieStore(nsIGlobalObject* aGlobal)
     : DOMEventTargetHelper(aGlobal) {
   mNotifier = CookieStoreNotifier::Create(this);
+
+  // This must be created _after_ CookieStoreNotifier because we rely on the
+  // notification order.
+  mNotificationWatcher = CookieStoreNotificationWatcherWrapper::Create(this);
 }
 
 CookieStore::~CookieStore() { Shutdown(); }
@@ -321,6 +326,21 @@
           return;
         }
 
+        if (!self->mNotificationWatcher) {
+          promise->MaybeReject(NS_ERROR_UNEXPECTED);
+          return;
+        }
+
+        nsID operationID;
+        rv = nsID::GenerateUUIDInPlace(operationID);
+        if (NS_WARN_IF(NS_FAILED(rv))) {
+          promise->MaybeReject(NS_ERROR_UNEXPECTED);
+          return;
+        }
+
+        self->mNotificationWatcher->ResolvePromiseWhenNotified(operationID,
+                                                               promise);
+
         RefPtr<CookieStoreChild::SetRequestPromise> ipcPromise =
             self->mActor->SendSetRequest(
                 aOptions.mDomain.IsEmpty() ? nsString(baseDomain)
@@ -333,7 +353,7 @@
                     ? INT64_MAX
                     : static_cast<int64_t>(aOptions.mExpires.Value() / 1000),
                 path, SameSiteToConst(aOptions.mSameSite),
-                aOptions.mPartitioned);
+                aOptions.mPartitioned, operationID);
         if (NS_WARN_IF(!ipcPromise)) {
           promise->MaybeResolveWithUndefined();
           return;
@@ -341,9 +361,15 @@
 
         ipcPromise->Then(
             NS_GetCurrentThread(), __func__,
-            [promise = RefPtr<dom::Promise>(promise)](
+            [promise = RefPtr<dom::Promise>(promise), self = RefPtr(self),
+             operationID](
                 const CookieStoreChild::SetRequestPromise::ResolveOrRejectValue&
-                    aResult) { promise->MaybeResolveWithUndefined(); });
+                    aResult) {
+              if (!aResult.ResolveValue()) {
+                self->mNotificationWatcher->ForgetOperationID(operationID);
+                promise->MaybeResolveWithUndefined();
+              }
+            });
       }));
 
   return promise.forget();
@@ -410,24 +436,44 @@
           return;
         }
 
+        if (!self->mNotificationWatcher) {
+          promise->MaybeReject(NS_ERROR_UNEXPECTED);
+          return;
+        }
+
+        nsID operationID;
+        rv = nsID::GenerateUUIDInPlace(operationID);
+        if (NS_WARN_IF(NS_FAILED(rv))) {
+          promise->MaybeReject(NS_ERROR_UNEXPECTED);
+          return;
+        }
+
+        self->mNotificationWatcher->ResolvePromiseWhenNotified(operationID,
+                                                               promise);
+
         RefPtr<CookieStoreChild::DeleteRequestPromise> ipcPromise =
             self->mActor->SendDeleteRequest(
                 aOptions.mDomain.IsEmpty() ? nsString(baseDomain)
                                            : nsString(aOptions.mDomain),
                 cookiePrincipal->OriginAttributesRef(),
-                nsString(aOptions.mName), path, aOptions.mPartitioned);
+                nsString(aOptions.mName), path, aOptions.mPartitioned,
+                operationID);
         if (NS_WARN_IF(!ipcPromise)) {
           promise->MaybeResolveWithUndefined();
           return;
         }
 
-        ipcPromise->Then(NS_GetCurrentThread(), __func__,
-                         [promise = RefPtr<dom::Promise>(promise)](
-                             const CookieStoreChild::DeleteRequestPromise::
-                                 ResolveOrRejectValue& aResult) {
-                           MOZ_ASSERT(aResult.IsResolve());
-                           promise->MaybeResolveWithUndefined();
-                         });
+        ipcPromise->Then(
+            NS_GetCurrentThread(), __func__,
+            [promise = RefPtr<dom::Promise>(promise), self = RefPtr(self),
+             operationID](const CookieStoreChild::DeleteRequestPromise::
+                              ResolveOrRejectValue& aResult) {
+              MOZ_ASSERT(aResult.IsResolve());
+              if (!aResult.ResolveValue()) {
+                self->mNotificationWatcher->ForgetOperationID(operationID);
+                promise->MaybeResolveWithUndefined();
+              }
+            });
       }));
 
   return promise.forget();